Introduce CUDA Device Assertions Infrastructure

r-barnes opened the pull request:
Summary:
This diff introduces a set of changes that makes it possible for the host to get assertions from CUDA devices. This includes the introduction of

**`CUDA_KERNEL_ASSERT2`**

A preprocessor macro to be used within a CUDA kernel that, upon an assertion failure, writes the assertion message, file, line number, and possibly other information to UVM (Managed memory). Once this is done, the original assertion is triggered, which places the GPU in a Bad State requiring recovery. In my tests, data written to UVM appears there before the GPU reaches the Bad State and is still accessible from the host after the GPU is in this state.

Messages are written to a multi-message buffer which can, in theory, hold many assertion failures. I've done this as a precaution in case there are several, but I don't actually know whether that is possible and a simpler design which holds only a single message may well be all that is necessary.

**`TORCH_DSA_KERNEL_ARGS`**

This preprocess macro is added as an _argument_ to a kernel function's signature. It expands to supply the standardized names of all the arguments needed by `C10_CUDA_COMMUNICATING_KERNEL_ASSERTION` to handle device-side assertions. This includes, eg, the name of the pointer to the UVM memory the assertion would be written to. This macro abstracts the arguments so there is a single point of change if the system needs to be modified.

**`c10::cuda::get_global_cuda_kernel_launch_registry()`**

This host-side function returns a singleton object that manages the host's part of the device-side assertions. Upon allocation, the singleton allocates sufficient UVM (Managed) memory to hold information about several device-side assertion failures. The singleton also provides methods for getting the current traceback (used to identify when a kernel was launched). To avoid consuming all the host's memory the singleton stores launches in a circular buffer; a unique "generation number" is used to ensure that kernel launch failures map to their actual launch points (in the case that the circular buffer wraps before the failure is detected).

**`TORCH_DSA_KERNEL_LAUNCH`**

This host-side preprocessor macro replaces the standard
```
kernel_name<<<blocks, threads, shmem, stream>>>(args)
```
invocation with
```
TORCH_DSA_KERNEL_LAUNCH(blocks, threads, shmem, stream, args);
```
Internally, it fetches the UVM (Managed) pointer and generation number from the singleton and append these to the standard argument list. It also checks to ensure the kernel launches correctly. This abstraction on kernel launches can be modified to provide additional safety/logging.

**`c10::cuda::c10_retrieve_device_side_assertion_info`**
This host-side function checks, when called, that no kernel assertions have occurred. If one has. It then raises an exception with:
1. Information (file, line number) of what kernel was launched.
2. Information (file, line number, message) about the device-side assertion
3. Information (file, line number) about where the failure was detected.

**Checking for device-side assertions**

Device-side assertions are most likely to be noticed by the host when a CUDA API call such as `cudaDeviceSynchronize` is made and fails with a `cudaError_t` indicating
> CUDA error: device-side assert triggered CUDA kernel errors

Therefore, we rewrite `C10_CUDA_CHECK()` to include a call to `c10_retrieve_device_side_assertion_info()`. To make the code cleaner, most of the logic of `C10_CUDA_CHECK()` is now contained within a new function `c10_cuda_check_implementation()` to which `C10_CUDA_CHECK` passes the preprocessor information about filenames, function names, and line numbers. (In C++20 we can use `std::source_location` to eliminate macros entirely!)

# Notes on special cases

* Multiple assertions from the same block are recorded
* Multiple assertions from different blocks are recorded
* Launching kernels from many threads on many streams seems to be handled correctly
* If two process are using the same GPU and one of the processes fails with a device-side assertion the other process continues without issue
* X Multiple assertions from separate kernels on different streams seem to be recorded, but we can't reproduce the test condition
* X Multiple assertions from separate devices should be all be shown upon exit, but we've been unable to generate a test that produces this condition

Differential Revision: D37621532

soumith commented:
this is pretty cool!

r-barnes commented:
Failures in "pull / linux-bionic-cuda11.6-py3.10-gcc7-bazel-test / build-and-test" look unrelated.

dzhulgakov commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,101 @@
+          blockIdx,                                                         \
+          threadIdx);                                                       \
+                                                                            \
+      assert(condition);                                                    \
```
is it ok that we evaluate condition twice? it shouldn't have side effects but still. Maybe call __assert_fail directly?

r-barnes replied:
It depends. If we call it twice we can both capture the assertion failure ourselves while still providing CUDA's standard assertion failure handling. If we `__assert_fail` directly, then we cannot provide such handling.

r-barnes replied:
I've gone with `__assert_fail`.

dzhulgakov commented on c10/cuda/CUDAException.cpp:
```diff
@@ -1,13 +1,96 @@
+    oss << failures_found
+        << " CUDA device-side assertion failures were found on GPU #"
+        << device_num << "!" << std::endl;
+    oss << "But at least " << assertion_data_for_device.assertion_count
```
should this be included only if number of asserts is more than C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT?

r-barnes replied:
I've incorporated your suggestion, thanks.

dzhulgakov commented on c10/cuda/CUDAException.cpp:
```diff
@@ -26,10 +109,192 @@ void c10_cuda_check_implementation(
+      launch_filename,
+      launch_function,
+      launch_linenum,
+      c10::get_backtrace(),
```
how slow is getting backtrace? maybe make it opt-in? everything else should be quick. If you want to keep it - move it outside of the mutex

r-barnes replied:
I've added a `TORCH_CUDA_DSA_WITH_BACKTRACE` preprocessor gate (unless you think it'd be better to have this be opt-in at runtime?) and moved collection outside of the mutex.

dzhulgakov replied:
hm, it's probably better to make it a runtime opt-in. Environment variable perhaps? Users definitely won't recompile torch to debug things. I'd also make the whole reporting controllable by an env variable in addition to the preprocessor flag. So you can have settings like off/without_stack/with_stack

dzhulgakov commented on c10/cuda/CUDAException.cpp:
```diff
@@ -26,10 +109,192 @@ void c10_cuda_check_implementation(
+  // If we've already set up this GPU with managed memory, return a pointer to
+  // the managed memory
+  if (uvm_assertions.at(device_num)) {
+    std::cerr << "Creating a new uvm_assertions pointer" << std::endl;
```
this is not a failure situation - why spam the output?

r-barnes replied:
I've removed this line - thanks!

dzhulgakov reviewed (commented):
This is really nice! Did you run it through microbenchmarks to see the launch overhead?

ezyang commented:
added @andreaskoepf to the reviewer list, he did an original version of this idea at https://github.com/pytorch/pytorch/pull/26776 ; maybe this time we can actually land it!

r-barnes commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,105 @@
+          blockIdx,                                                         \
+          threadIdx);                                                       \
+                                                                            \
+      assert(condition);                                                    \
```
A design that evaluates `condition` once would call `assert(condition_result)` instead of `assert(condition)`, but `assert(condition_result)` would cause CUDA to print a less informative error message. That might be okay since our intention is to capture the error message anyway, but it ultimately reduces flexibility for users without much gain since, as you say, the assertions should (hopefully) be without side-effects.

> But perhaps you could call some internals like __assert_fail to provide an explicit message?

I'm not sure what you mean by this.

dzhulgakov replied:
`assert` is usually implemented as a macro somewhere inside CUDA. So you could depend on implementation detail and call the function that the macro calls underneath. In std c library it's usually `__assert_fail` (e.g. https://github.com/lattera/glibc/blob/master/assert/assert.h#L67)

r-barnes replied:
I've gone with __assert_fail.

andreaskoepf commented:
@r-barnes, great to see someone working on cuda asserts again!

When I worked on the graceful-assert feature in https://github.com/pytorch/pytorch/pull/26776 I noticed negative performance impacts as soon as (too many) string-literals appeared in the kernel (either the assert message or `__FILE__` macros), e.g. see my comments  https://github.com/pytorch/pytorch/pull/26776#issuecomment-537575482 and https://github.com/pytorch/pytorch/pull/26776#issuecomment-537854481. I would therefore recommend to run some micro-benchmarks to measure and compare performance of realistic kernels with and without the new cuda-assert infrastructure. In the best case the cuda compiler has been improved during the last years or what I noticed back then simply does not apply to the new approach.

General question: Are there now ways to recover from sticky cuda errors (e.g. `assert(false)`) without terminating the host-process, e.g. is `cudaDeviceReset(`) now enough?

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html says:
> cudaErrorAssert = 710
>     An assert triggered in device code during kernel execution. The device cannot be used again. All existing allocations are invalid. To continue using CUDA, the process must be terminated and relaunched. 
> 

but https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html says:

> Any subsequent host-side synchronization calls made for the same device will return cudaErrorAssert. No more commands can be sent to this device until cudaDeviceReset() is called to reinitialize the device.

One of my main motivations in 2019 was to gracefully handle assert-failures in kernels .. I approached it with graceful-returns from the kernels but proper device recovery would of course be much better.

r-barnes commented:
@andreaskoepf - Are there existing benchmarks I can use to test for performance degradation?

Device-side assertions still cannot be recovered from without killing the associated host process and this setup does not attempt to work around that. The main goal here is to capture the failed assertions from the device, associate them with whatever kernel launch triggered the failure, and package all this information neatly in one place to facilitate easier debugging.

r-barnes commented:
@andreaskoepf - Thanks! When you say
> The module level output when compiling MultinomialKernel.cu with --resource-usage

I wonder how you did that compilation to get the ptxas info?

r-barnes commented:
@andreaskoepf

I've built a benchmark on the branch [here](https://github.com/r-barnes/pytorch/blob/dsa_benchmark/temp/dsa_benchmark.cpp), which I've compiled with
```
reset && time CXX=clang++-10 CC=clang-10 USE_CUDA=1 USE_ROCM=0 TORCH_CUDA_ARCH_LIST="8.0" USE_KINETO=0 BUILD_TEST=0 USE_DISTRIBUTED=0 BUILD_CAFFE2_OPS=0 USE_FBGEMM=0 BUILD_CAFFE2=0 USE_OPENCV=0 USE_QNNPACK=0 USE_XNNPACK=0 DEBUG=0 USE_GOLD_LINKER=1 CCACHE_DISABLE=1 python setup.py develop 
```
A diff between the benchmark and this DSA (device-side assertion) infrastructure branch is [here](https://github.com/r-barnes/pytorch/compare/export-D37621532...r-barnes:pytorch:dsa_benchmark).

This uses the same multinomial code as your benchmark, but I've kicked it up to 1k iterations because the N=10 sample size seemed like it was mostly producing noise.

We get the following results:
```
# Baseline (1000 trials per test)
Test #1 2071088
Test #2 2074427
Test #3 2076612
Test #4 2080450
Test #5 2084301
Test #6 2094876
Test #7 2097009
Test #8 2105669
std=11.5292313707376 μs
avg=2085.554 μs

# With DSA (1000 trials per test)
Test #1 2062365
Test #2 2063063
Test #3 2064451
Test #4 2068551
Test #5 2069857
Test #6 2077091
Test #7 2097987
Test #8 2301278
std=76.63315184849424 μs 
avg=2100.580375 μs
```

The standard deviation is 65μs higher with DSA, but I wouldn't expect anyone to notice that. The time difference between the two, at 15μs also seems pretty small.

ezyang commented:
What CUDA version is the benchmark on?

r-barnes commented:
@ezyang - This is CUDA v11.6.112 (I've edited the description above with this info as well).

ezyang commented:
It's possible that @andreaskopf's cited slowdown was specific to CUDA 10. If so, happy day; since we're dropping 10.2 support. If you want to be totally sure, test with the earliest CUDA version we support (sorry, I actually don't remember which point release it is...)

ezyang commented:
11.3, according to @janeyx99 !

I wonder if the fact that we no longer need out of line string literals means we can also stop trashing the CUDA context. Maybe we can do it as follow up.

janeyx99 commented:
@r-barnes @ezyang it appears 11.3 was a misleading answer! the stable torch 1.12 still supports 11.3 but our nightly binaries have 11.6 as the lowest cuda version we support.

andreaskoepf commented:
Probably the macro as it is currently proposed `CUDA_KERNEL_ASSERT2(condition)` is fine .. developers would have to write something like `&& "msg"` to add string messages and probably most of the time will not naturally do this. For debug purposes it is already great to get a string-representation of the failing condition together with the source-reference and at least in the multinomial test that seemed not to harm performance. Overall 👍 from my side.

r-barnes commented:
@andreaskoepf - Apologies, I've added print statements to both the CUDA and CPU multinomial code and it doesn't look as though my benchmark is using either. I'm not sure what's going wrong, but it looks like the code is getting its multinomial implementation from some place other than the source I modified.

andreaskoepf commented:
> I wonder how you did that compilation to get the ptxas info?

I probably added [--resource-usage (-res-usage)](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#generic-tool-options-resource-usage) or `--ptxas-options=-v` to the nvcc args.

ezyang commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,106 @@
+namespace c10 {
+namespace cuda {
+
+#ifdef TORCH_USE_CUDA_DSA
```
Err, do we have to feature gate this? Would be better if we can just turn this on unconditionally everywhere.

r-barnes replied:
@ezyang  Having the feature gate is useful for disabling this feature on older Nvidia GPUs and most AMD GPUs. The older Nvidia GPUs might not be an issue, but the AMD GPUs might. We can try without the gating, if you prefer.

ezyang commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,106 @@
+  }
+}
+
+__device__ __noinline__ void dsa_add_new_assertion_failure(
```
since this is noinline, I wonder if we can avoid putting its implementation in a header file

ezyang commented on c10/cuda/CUDAException.cpp:
```diff
@@ -26,10 +111,197 @@ void c10_cuda_check_implementation(
+  int device_count = -1;
+  C10_CUDA_ERROR_HANDLED(cudaGetDeviceCount(&device_count));
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false);
+  return device_count;
```
It's not clear why this has to be a method on the class

dzhulgakov replied:
can't you call `c10::cuda::device_count/current_device()` directly? There's a circular dependency between the headers, but since in .cpp file it should be ok.

I'd avoid duplicating the calls.

ezyang replied:
Sure, but it still doesn't have to be a method on the class :P

dzhulgakov commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,106 @@
+#include <c10/macros/Macros.h>
+
+// Used by preprocessor to convert things to strings
+#define C10_CUDA_DEVICE_ASSERTION_STRINGIFY(x) #x
```
There's `C10_STRINGIZE` already

dzhulgakov commented on c10/cuda/CUDAException.cpp:
```diff
@@ -26,10 +111,197 @@ void c10_cuda_check_implementation(
+    const std::string& kernel_name,
+    const int32_t stream_id) {
+#ifdef TORCH_USE_CUDA_DSA
+#ifdef TORCH_CUDA_DSA_WITH_BACKTRACE
```
Repeating the comment: can we make stack control a runtime opt-in instead? It doesn't have any static impact on the performance. It can be an env variable perhaps.

Also as Ed mentioned if there's little perf impact from enabling the logging by default - we should make `TORCH_USE_CUDA_DSA` a runtime option too.

r-barnes replied:
I have added runtime options for both stacktrace acquisition and DSA as a whole.

ezyang commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,114 @@
+    // Therefore, we delay until those threads are done writing.
+    while (assertions_data->assertion_failure_written <
+           assertions_data->assertion_count) {
+    }
```
Wondering if we should have a timeout here 🤔

ezyang replied:
So, as @andreaskoepf mentioned, in his original implementation, he doesn't bother trapping; instead there's a boolean field that we regularly check on kernel launch that is stand-in for "an assert has failed". The big benefit is that we don't trash the CUDA context on an assertion fail, which is nice for interactive usage. This would also have the side effect of solving the race condition that you're having the threads busy wait for.

Even if you don't actually implement this, it seems like you could just have the thread fall through and keep running the rest of the kernel. The only problem is to make sure you have a "safe" value for indexing, but as mentioned, this seems like we ought to do anyway.

r-barnes replied:
I'll give this a try.

r-barnes replied:
@ezyang : The problem is: how do you communicate to the host that something's gone wrong on the GPU?

Raising an assert on the GPU triggers a failure mode that gets caught by `C10_CHECK` on any API call or kernel launch. So the host recognizes there was an issue on the next interaction with the GPU.

Raising a flag requires the host to check UVM at each API call and kernel launch to achieve the same behaviour. If we model the GPU's write to UVM as expensive (due device-host communication) but occurring only when an assert fails and the host's read from UVM as being cheap, then this is probably fine.

ezyang replied:
Andreas had some gadget for asynchronously communicating back to host, so the host tests were cheap and just delayed.

However, @ngimel made the case that delaying the errors is not exactly the best since the tensors are trashed. So maybe delaying as you did here is reasonable.

r-barnes replied:
Earlier I was hitting a problem where:
* `assert()` wasn't always causing the GPU to go into failed device mode reliably
* `__trap()` always worked but raises an exception in the host process
* The above seemed fragile

So making a mark in UVM memory to indicate failure and then having the threads quietly produce garbage seems fine as a way of working around the above and trying to preserve the host process.

I've looked at Andreas's code, but I haven't found the gadget. I'm assuming it'll be a write to memory by the device accompanied by checking said memory on the host.

r-barnes replied:
Hmmm, if the `assert` was meant to protect against, eg, an out-of-bounds array access, then continuing on through the kernel is the wrong action and an early exit is needed.

r-barnes replied:
@andreaskoepf, maybe @ngimel has more thoughts, but if there's a situation where we can perform out-of-bounds memory accesses and use `assert()` to stop execution before this, that's a situation where trash-and-check may provide inadequate safeguards.

I've observed that the hard `assert` seems to sometimes get skipped in ways I can't reliably reproduce, so I'm switching to the trash-and-check strategy anyway.

r-barnes replied:
I've now switched this to not using `assert()` at all, trying to early-exit assert-failed kernels where possible, and relying solely on UVM to communicate assertion failures.

ezyang commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
+ private:
+  /// Assume that this is the max number of kernel launches that might ever be
+  /// enqueued across all streams on a single device
+  static constexpr int max_kernel_launches = 1024;
```
How did we compute this number? What happens if we violate this constraint?

r-barnes replied:
We chose this out of a hat. If we violate this constraint the circular buffer wraps around on itself. No error occurs and the generation/launch id ensures that we still only present the user with valid debugging info.

ezyang commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  /// Register a new kernel launch and obtain a generation number back to be
+  /// passed to the kernel
+  uint32_t insert(
+      const std::string& launch_filename,
```
It seems unnecessary for these to be std::string; you're always passing in `__FILE__` so these can just be `const char*`. It seems good to apply this optimization, as you are ALWAYS populating these fields for every DSA cuda launch, seems much better to not have to do a dynamic allocation every time you do this.

r-barnes replied:
Good call.

ezyang commented:
Sorry, was really behind on reviewing this.

At a high level, this looks pretty reasonable. I am primarily worried about the performance implications of the patch. It would be good to hook this up to at least one real kernel in this PR, and then microbenchmark the impact, including in multithreaded settings. I'd also love to know if we can avoid trashing the cuda context, this matters a lot to folks and if inline strings are OK for perf, I don't think it's that much more work to do.

ngimel commented:
@ezyang wdym not trashing cuda context?

ezyang commented:
If you trigger a cuda assert in a notebook you don't have to restart the notebook

ezyang commented:
ok fine, suggestion retracted

r-barnes commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  /// Line number of where the code was launched from
+  uint32_t launch_linenum;
+  /// Backtrace of where the kernel was launched from
+  std::string launch_stacktrace;
```
We have a run-time flag that enables/disables this behaviour. It is disabled by default.

r-barnes replied:
Added comment clarifying this.

r-barnes commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  /// mark where we are inserting into the queue
+  uint32_t generation_number = 0;
+  /// Shared mutex between writer and accessor to ensure multi-threaded safety.
+  mutable std::mutex read_write_mutex;
```
Good idea, thanks.

r-barnes replied:
Hmm, but if we do this then it might be that one thread's kernel has an assertion failure and then freezes. A second thread detects that the assertion failure has happened, but can't report on it because it does not have access to the first thread's `kernel_launches` vector.

ezyang replied:
Hmm, I'm not sure I understand. Why does the kernel launched on the other thread need to know about an unrelated kernel that failed?

r-barnes replied:
I think the alternative is making the communication between the GPU and CPU more complex or expensive.

ezyang replied:
I guess my point is, if we're crashing the cuda context, it doesn't really matter what happens in the other threads; we want to get the errors out and then crash the process.

r-barnes replied:
@ezyang - That's true, but the thread that initiates the crash (by discovering there was a problem) may not be the thread that has the kernel launch information associated with the failure.

This mutex is here because, iirc, we were worried that making `generation_number` atomic wouldn't be sufficient to prevent fast wrapping on the launch buffer.

I'd guess that multithreaded (as opposed to multiprocess) GPU handling is probably an edge case, so both the mutex and atomic will be low-contention.

ezyang replied:
No, I don't think this is accurate. Things like torchdeploy and also C++ API are specifically for letting us do inferences while multithreading on one process.

r-barnes replied:
The benchmark below shows that the mutex doesn't affect speed.

Thinking about it more, if each thread allocated it's own assertion failure buffer on each GPU it interacted with, that would also solve the problem and eliminate the mutex. But that would also use (on the machine I'm working on) 80x more memory.

The current on-device memory usage is ~15kB. 80x that would be 1.2MB, which is still probably just fine on today's multigigabyte GPUs.

Still, it might be better to get this landed so folks can begin using it and switch to separate buffers in a follow-up.

r-barnes commented:
@ezyang: We can save the CUDA context, but it will be difficult to truly recover from a device-side assertion failure without rolling back to some known checkpoint. CUDA kernels run asynchronously, which means _many_ kernels can be scheduled to run without intervening host operations that could detect device-side failures. Once a device-side failure has been detected it would still be difficult to determine which kernels ran before the failure occurred. 

`C10_CUDA_CHECK(cudaDeviceSynchronize())` is the only call we can make that's immune to this issue since`cudaDeviceSynchronize` blcoks until all kernels have run and then `C10_CUDA_CHECK` can check for failures. If this succeeds, then the code is at a known good point. If this fails, then some tensors have been garbagified since the last such call.

ezyang commented:
I think the most practical way to figure out if the tensors are valid or not would to be have some globally increasing atomic sequence number that we assign to kernel launches. When an assert fails we note which sequence number failed, and then all tensors after that sequence number are now poisoned.

r-barnes commented:
@ezyang - That counter's already in this PR. But I don't think there's a convenient way to map kernel launches to their associated (Python?) output tensors.

ezyang commented:
Yes, we would have to have space to store this on the tensor somehow

r-barnes commented:
@ezyang - I suggest we leave that follow-up work, then.

r-barnes commented:
Works, but the perf is bad. I need to do some benchmarking tomorrow and figure out where the time is getting spent:
```
 100347578 - long strings
  93703317 - no strings
   6090589 - DSA disabled at compile time
```

r-barnes commented:
@andreaskoepf @ezyang : I've just finished updating the code to incorporate your comments and the benchmarks (definitely working this time) are looking good. What do we still need to do to move this forward?

ezyang commented:
I think my biggest problem is the mutex impact on multithreaded performance. Otherwise the singlethreaded perf numbers are good enough for me to give it a spin.

r-barnes commented:
@ezyang - Options:
* Further reductions in mutex scope (in particular, moving stacktrace generation outside the mutex)
* `thread_local`, though I don't like this for the reasons mentioned above
* Make `generation_number` an atomic, but possibly encountering race conditions if the buffer wraps too quickly
* Per-thread buffers, but managed via a mechanism other than `thread_local`
* Interthread communication following an assertion failure to fetch `thread_local` data

Do you have other ideas?

r-barnes commented:
Using an adaptation of the [multinomial benchmark](https://github.com/r-barnes/pytorch/blob/dsa_benchmark/temp/dsa_benchmark.cpp) that runs 20 threads simultaneously, I observe:
```
DSA: Mulithreaded, long-strings, 1024x1024 
Total: 125814612 us
Per kernel: 6.3ms

DSA: Multithreaded, no strings, 1024x1024 
Measuring...
Total: 116676131 us
Per kernel: 5.8ms

*no* DSA: Multithreaded, no strings, 1024x1024 
Total: 116276542 us
Per kernel: 5.8ms

DSA: Mulithreaded, no strings, 128x128
Total: 2630068 us
Per kernel: 0.13ms

*no* DSA: Mulithreaded, no strings, 128x128
Total: 3247621 us
Per kernel: 0.16ms
```
So the mutex doesn't seem to be materially affecting speed for either large kernels (where you might expect the cost to be amortized) or small kernels (where you might not).

ezyang reviewed (approved):
yolo

malfet commented on c10/cuda/test/impl/CUDAAssertionsTest.cu:
```diff
@@ -0,0 +1,658 @@
+#include <c10/cuda/CUDAFunctions.h>
+#include <c10/cuda/CUDAStream.h>
+
+#include <sys/wait.h>
```
Is this a Linux specific test? As there are no `sys/wait.h` on Windows

r-barnes replied:
@malfet - Apologies; I can't believe I missed that. I've removed the header and am now retesting.

malfet reviewed (changes requested):
Your PR fails to compile on windows with:
```
C:\actions-runner\_work\pytorch\pytorch\c10\cuda\test\impl\CUDAAssertionsTest.cu(9): fatal error C1083: Cannot open include file: 'sys/wait.h': No such file or directory
CUDAAssertionsTest.cu
```

malfet commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,104 @@
+  // `dst` ends with a null-terminator from `src`. Otherwise, the loop
+  // reached maximum iterations and we add our own null-terminator to
+  // `dst`
+  if (i == C10_CUDA_DSA_MAX_STR_LEN - 1) {
+    *dst = '\0';
+  }
```
It will set a null at the beginning of the string in that case, isn't it?

r-barnes replied:
Good catch.

malfet commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,104 @@
+
+#ifdef TORCH_USE_CUDA_DSA
+// Copy string from `src` to `dst`
+static __device__ void dstrcpy(char* const dst, const char* const src) {
```
Not sure if it supported across all cuda versions, but isn't it just `strncpy`? (and if it is not, perhaps it should be called `strncpy`

Also, not sure if `static` keyword has any effect on `__device__` only functions (as they do not have global visibility to beging with)

r-barnes replied:
Leaving the `static` since my recollection is someone else requested it earlier in review and it seems otherwise benign.

I don't think `strncpy` works in CUDA and am not sure if we can adequately test for that, so I'd prefer to keep this for now.

This function's signature differs from `strncpy`, so I'd prefer to keep the name as-is unless you have other suggestions.

tl;dr I haven't made a change here.

malfet commented on c10/cuda/CUDADeviceAssertion.h:
```diff
@@ -0,0 +1,104 @@
+  int i = 0;
+  // Copy string from source to destination, ensuring that it
+  // isn't longer than `C10_CUDA_DSA_MAX_STR_LEN-1`
+  for (; src[i] != '\0' && i < C10_CUDA_DSA_MAX_STR_LEN - 1; i++) {
+    dst[i] = src[i];
+  }
+  // If the loop terminated before `i` meets the condition below, then
+  // `dst` ends with a null-terminator from `src`. Otherwise, the loop
+  // reached maximum iterations and we add our own null-terminator to
+  // `dst`
+  if (i == C10_CUDA_DSA_MAX_STR_LEN - 1) {
+    *dst = '\0';
+  }
```
This sounds like a while loop to me, that copies up to C10_CUDA_DSA_MAX_STR_LEN-1 characters from src to dst until `\0` is met
```suggestion
  while ( *src != '\0' && i++ < C10_CUDA_DSA_MAX_STR_LEN - 1) {
    *dst++ = *src++;
  }
  *dst = '\0';
```

r-barnes replied:
I've updated the code to adopt this suggestion.

malfet commented on c10/cuda/CUDAException.cpp:
```diff
@@ -1,35 +1,380 @@
+        C10_CUDA_DSA_ASSERTION_COUNT) {
+      oss << "But at least " << assertion_data_for_device.assertion_count
+          << " assertion failures occurred on the device" << std::endl;
+      oss << "Adjust `C10_CUDA_DSA_ASSERTION_COUNT` if you need more assertion failure info"
```
Though kinds of messages often get out of date (as one edits the environment variable but ignores the rest

r-barnes replied:
Do you have a proposed solution?

malfet commented on c10/cuda/CUDAException.h:
```diff
@@ -70,6 +80,13 @@ class C10_CUDA_API CUDAError : public c10::Error {
+/// threads will fail silently.
+// We use a preprocessor macro here so the value will also be available in
+// the CUDA code.
+#define C10_CUDA_DSA_ASSERTION_COUNT 10
+#define C10_CUDA_DSA_MAX_STR_LEN 512
```
Why those needs to be preprocessor macros rather than constexprs?

r-barnes replied:
I don't think it'll work with `constexpr`, but we can try.

r-barnes replied:
Looks like it worked. Nice catch.

r-barnes commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +98,168 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  /// A single circular buffer holds information about every kernel launch the
+  /// process makes across all devices.
+  std::vector<CUDAKernelLaunchInfo> kernel_launches;
+  /// Get the number of CUDA devices
+  static int get_device_count();
+  /// Get current device id - a little redundant, but helps avoid a circular
+  /// include
+  static int get_device_id();
+  /// Get current device compute capability - note that this dangerously assumes
+  /// that if one CUDA GPU supports device-side assertions they all do. This is
+  /// probably fine since the latest CUDA GPU that doesn't support UVM is the
+  /// K80 released 2014-11-17. Mixing that GPU with a newer one is likely to be
+  /// rare enough that the defensive
+  static int get_device_compute_capability(const int device_num);
+  static bool check_if_all_devices_support_managed_memory();
```
However, the functions are _very_ specific to device-side assertions since, internally, they override safety checks that would cause an infinite initialization loop for the `CUDAKernelLaunchRegistry` singleton.

r-barnes commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +98,168 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  /// Get current device id - a little redundant, but helps avoid a circular
+  /// include
+  static int get_device_id();
+  /// Get current device compute capability - note that this dangerously assumes
+  /// that if one CUDA GPU supports device-side assertions they all do. This is
+  /// probably fine since the latest CUDA GPU that doesn't support UVM is the
+  /// K80 released 2014-11-17. Mixing that GPU with a newer one is likely to be
+  /// rare enough that the defensive
```
It'd be helpful if I knew which part you were referring to. I do note that the comment assumes the function applies only to a single device, though, and I've changed the comment to encompass the generality of the current implementation.

r-barnes commented on c10/cuda/CUDAException.h:
```diff
@@ -81,5 +98,168 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  DeviceAssertionsData* get_uvm_assertions_ptr_for_current_device();
+  /// Gets the global singleton of the registry
+  static CUDAKernelLaunchRegistry& get_singleton_ref();
+  /// Number of GPUs I've interacted with
+  int gpus_interacted_with() const;
```
I've removed the function and instead in this case print:
```
This process interacted the following GPUs = {0,2,4,5}
```

malfet commented on c10/cuda/CUDADeviceAssertionHost.cpp:
```diff
@@ -0,0 +1,366 @@
+#include <string>
+#include <thread>
+
+#define CHECK_CUDA_API_CALL_WITHOUT_CHECKING_DEVICE_ASSERTS \
```
Nit: can you please make it look like a function?
```suggestion
#define CHECK_CUDA_API_CALL_WITHOUT_CHECKING_DEVICE_ASSERTS() \
```

r-barnes replied:
Done, thanks.

r-barnes committed:
Introduce CUDA Device Assertions Infrastructure (#84609)

Summary:
This diff introduces a set of changes that makes it possible for the host to get assertions from CUDA devices. This includes the introduction of

**`CUDA_KERNEL_ASSERT2`**

A preprocessor macro to be used within a CUDA kernel that, upon an assertion failure, writes the assertion message, file, line number, and possibly other information to UVM (Managed memory). Once this is done, the original assertion is triggered, which places the GPU in a Bad State requiring recovery. In my tests, data written to UVM appears there before the GPU reaches the Bad State and is still accessible from the host after the GPU is in this state.

Messages are written to a multi-message buffer which can, in theory, hold many assertion failures. I've done this as a precaution in case there are several, but I don't actually know whether that is possible and a simpler design which holds only a single message may well be all that is necessary.

**`TORCH_DSA_KERNEL_ARGS`**

This preprocess macro is added as an _argument_ to a kernel function's signature. It expands to supply the standardized names of all the arguments needed by `C10_CUDA_COMMUNICATING_KERNEL_ASSERTION` to handle device-side assertions. This includes, eg, the name of the pointer to the UVM memory the assertion would be written to. This macro abstracts the arguments so there is a single point of change if the system needs to be modified.

**`c10::cuda::get_global_cuda_kernel_launch_registry()`**

This host-side function returns a singleton object that manages the host's part of the device-side assertions. Upon allocation, the singleton allocates sufficient UVM (Managed) memory to hold information about several device-side assertion failures. The singleton also provides methods for getting the current traceback (used to identify when a kernel was launched). To avoid consuming all the host's memory the singleton stores launches in a circular buffer; a unique "generation number" is used to ensure that kernel launch failures map to their actual launch points (in the case that the circular buffer wraps before the failure is detected).

**`TORCH_DSA_KERNEL_LAUNCH`**

This host-side preprocessor macro replaces the standard
```
kernel_name<<<blocks, threads, shmem, stream>>>(args)
```
invocation with
```
TORCH_DSA_KERNEL_LAUNCH(blocks, threads, shmem, stream, args);
```
Internally, it fetches the UVM (Managed) pointer and generation number from the singleton and append these to the standard argument list. It also checks to ensure the kernel launches correctly. This abstraction on kernel launches can be modified to provide additional safety/logging.

**`c10::cuda::c10_retrieve_device_side_assertion_info`**
This host-side function checks, when called, that no kernel assertions have occurred. If one has. It then raises an exception with:
1. Information (file, line number) of what kernel was launched.
2. Information (file, line number, message) about the device-side assertion
3. Information (file, line number) about where the failure was detected.

**Checking for device-side assertions**

Device-side assertions are most likely to be noticed by the host when a CUDA API call such as `cudaDeviceSynchronize` is made and fails with a `cudaError_t` indicating
> CUDA error: device-side assert triggered CUDA kernel errors

Therefore, we rewrite `C10_CUDA_CHECK()` to include a call to `c10_retrieve_device_side_assertion_info()`. To make the code cleaner, most of the logic of `C10_CUDA_CHECK()` is now contained within a new function `c10_cuda_check_implementation()` to which `C10_CUDA_CHECK` passes the preprocessor information about filenames, function names, and line numbers. (In C++20 we can use `std::source_location` to eliminate macros entirely!)

# Notes on special cases

* Multiple assertions from the same block are recorded
* Multiple assertions from different blocks are recorded
* Launching kernels from many threads on many streams seems to be handled correctly
* If two process are using the same GPU and one of the processes fails with a device-side assertion the other process continues without issue
* X Multiple assertions from separate kernels on different streams seem to be recorded, but we can't reproduce the test condition
* X Multiple assertions from separate devices should be all be shown upon exit, but we've been unable to generate a test that produces this condition

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84609

Reviewed By: ezyang

Differential Revision: D37621532

Pulled By: r-barnes

fbshipit-source-id: efdfc57d6a1fa6dadfe30e693157e4cc040c7191

r-barnes commented:
@pytorchbot merge -f "Internal changes is incorrect."

r-barnes commented:
@pytorchbot merge -f "No internal changes"

r-barnes commented:
@pytorchbot merge

ppwwyyxx commented:
Why is this feature behind a compile-time flag `TORCH_USE_CUDA_DSA` that's default disabled? Does it have a significant runtime overhead even when not enabled at runtime, but just built in compile-time ?

The release not enabling this flag by default means that most users are not benefiting from this great feature.
